### ISSUE TITLE
Lmdb size on Windows

### ISSUE THREAD

**Issue by AndreaLanfranchi** [ledgerwatch/erigon#1134]
#### System information

Turbo-Geth version: `2020.09.02-alpha`
OS & Version: Windows
Commit hash : [master@47f59e8fcd](https://github.com/ledgerwatch/turbo-geth/tree/47f59e8fcd805ad2a4f677f7dffb2ae9b0673bbc)

#### Expected behaviour

Lmdb file should grow as data gets in.

#### Actual behaviour

On Windows lmdb file is immediately allocated the size given by `--lmdb.mapSize` cli argument (which defaults to 2TB).
By consequence, with default settings, you need at least 4TB free storage space as the databases being allocated are one for chain data and the other for nodes. Needless to say the latter only requires few megabytes.

#### Steps to reproduce the behaviour

Run a windows instance.

**Comment by AndreaLanfranchi** [comment]
Related issue #1115

**Comment by AndreaLanfranchi** [comment]
Additional documentation
https://docs.microsoft.com/it-it/windows/win32/memory/creating-a-file-mapping-object?redirectedfrom=MSDN

Seems like the only workaround is to trap MDB_MAP_FULL

**Comment by AskAlexSharov** [comment]
Here is example how can it be handled in go: https://github.com/ledgerwatch/lmdb-go/blob/1f23ded7b4bbf8b8ec2c7168271a5153b89c2379/lmdb/example_test.go#L188

Also, we already have waitgroup inside kv_lmdb - means we can wait for read transactions finish (but can’t hold new read transactions from start).

**Comment by AskAlexSharov** [comment]
Here is how monero handles it in cpp: https://github.com/monero-project/monero/blob/c61b3f0eadd4ad17398356d941684bc7de58ebcf/src/lmdb/database.cpp#L129

**Comment by AndreaLanfranchi** [comment]
Came up with smth similar
https://github.com/torquem-ch/silkworm/blob/stage-senders/silkworm/db/chaindb.cpp#L147-L167

**Comment by AskAlexSharov** [comment]
@AndreaLanfranchi , with which value of setMapSize must start process?

**Comment by AndreaLanfranchi** [comment]
I actually use --lmdb.mapSize="50 GB" ... but is subjective.
I only need to test. The value of mapsize is directly related to how much space you have on disk for Windows.
Default value in TG is 2TB ... which means 4TB files (2TB for chain data and 2TB for nodes).
Should differentiate map size for nodes and for chaindata

**Comment by AskAlexSharov** [comment]
I mean - what if we start from 1Gb, then increase to 2Gb, then to 3Gb, then restart node, with which value need to start now?

**Comment by AndreaLanfranchi** [comment]
Not sure I understand.
You can actually start with any value ... even 1MB if you wish. 
I think starting with 10GB and an increase lot of 2GB every time map is full would be fine.

**Comment by AskAlexSharov** [comment]
and after server restart - start from 10GB again?

**Comment by AndreaLanfranchi** [comment]
I don't follow or maybe I am missing something.
I'll try to clarify. The problem with mapsize on windows happens when value of map_size is greater than the actual size of file on disk. So on very first start of TG it will try to create a file on disk which has a size equal to map_size.
On **restart** you _may_ even specify a value of map_size to 1KB but, as the file on disk is already present env opens the file with a map size equal to the size of the file. So you basically never "restart" from 10GB.
In summary ... the bahavior on Windows is this :
* if map_size > file_size then enlarge the file
* if map_size <= file_size then override map_size with file_size

This is what I am observing

**Comment by AndreaLanfranchi** [comment]
I must add an interesting experiment I just completed.
I started up new instance (sync from scratch) using `tg.exe --lmdb.mapSize="50 GB"` and let runf for some time.
As expected two database files both of 50GB size are created in datadir.  After a couple fo hours I succesfully imported headers up to block 9.6M. At that moment I aborted with ^C. Database closed properly.
I eventually restarted the instance but this time using `tg.exe --lmdb.mapSize="5 GB"` (note the different map size). tg started properly and detected "Most recent local header" correctly but from that moment on it only emitted "Looking for peers" log lines even though the number of connected peers was well above 20. No header download whatsoever for over 2 hours. Eventually I aborted with ^C. 
Finally I restarted tg with the very first mapSize (50GB) and in a matter of minutes new imported headers began to flow in.

**Comment by tjayrush** [comment]
I saw this same problem recently (`tg version 2021.01.2-alpha`).

TG was generating this warning:

```
WARN [01-19|23:07:59.636] Synchronisation failed, dropping peer    peer=9ac8df96e49bce26 err="retrieved hash chain is invalid: commit inserting bodies: mdb_cursor_put: MDB_MAP_FULL: Environment mapsize limit reached
```

On Linux with this command line:

```
./build/bin/tg --datadir /mnt/md0/turbo-geth --private.api.addr=127.0.0.1:9090 --verbosity 3 --port 30304
```

I added `--lmdb.mapSize="4 TB" to the command line, and the warning seems to have disappeared.

**Comment by AskAlexSharov** [comment]
implemented by migration to mdbx
